### PR TITLE
[CDF-21456] No more magic

### PR DESCRIPTION
**Pull request by doctrino**
# Description

Please describe the change you have made.

## Checklist

- [ ] Tests added/updated.
- [ ] Run Demo Job Locally.
- [ ] Documentation updated.
- [ ] Changelogs updated in [CHANGELOG.cdf-tk.md](https://github.com/cognitedata/cdf-project-templates/blob/main/CHANGELOG.cdf-tk.md).
- [ ] Template changelogs updated in [CHANGELOG.templates.md](https://github.com/cognitedata/cdf-project-templates/blob/main/CHANGELOG.templates.md).
- [ ] Version bumped.
  [_version.py](https://github.com/cognitedata/cdf-project-templates/blob/main/cognite/cognite_toolkit/_version.py) and
  [pyproject.toml](https://github.com/cognitedata/cdf-project-templates/blob/main/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite_toolkit/_cdf_tk/load/_base_loaders.py [hunk 1]
```diff
@@ -243,34 +222,21 @@ def dump_resource(
...
-        elif isinstance(retrieved, Collection):
-            return self.list_cls(retrieved)
-        else:
-            return self.list_cls([retrieved])
+        raise NotImplementedError()
 
+    @abstractmethod
     def update(self, items: T_CogniteResourceList) -> Sized:
-        return self.api_class.update(items)
+        raise NotImplementedError()
```
Comment (ronpal) [inline cognite_toolkit/_cdf_tk/load/_base_loaders.py:235]: How about something like this, as a reminder? Or is it obvious when it is @abstractmethod?

```suggestion
        raise NotImplementedError("update must be implemented in all Resource Loader classes")

```

File: cognite_toolkit/_cdf_tk/load/_base_loaders.py [hunk 2]
```diff
@@ -243,34 +222,21 @@ def dump_resource(
...
 
+    @abstractmethod
     def delete(self, ids: SequenceNotStr[T_ID]) -> int:
-        existing = self.retrieve(ids)
-        if existing:
-            self.api_class.delete(ids)
-        return len(existing)
+        raise NotImplementedError()
```
Comment (ronpal) [inline cognite_toolkit/_cdf_tk/load/_base_loaders.py:239]:
```suggestion
        raise NotImplementedError("delete must be implemented in all Resource Loader classes")
```

**Review by ronpal — APPROVED** [review]
Approved, with comment/question

File: cognite_toolkit/_cdf_tk/load/_base_loaders.py (hunk 1 above)
Comment (doctrino) [reply cognite_toolkit/_cdf_tk/load/_base_loaders.py]: It should be obvious, https://stackoverflow.com/questions/40894284/what-should-i-put-in-the-body-of-an-abstract-method

Raising a NotImplementedError in abstractmethod means it should absolutely be implemented.

**Pull request merged**